### ISSUE TITLE
In-place functions

### ISSUE THREAD

**Issue by beam2d** [chainer/chainer#2353]
**tl;dr** Naive implementation of in-place functions cause bugs in user codes that are difficult to fix. We still have a way to implement it safely, which requires updates on existing codes.

In-place function is a function (I mean, a subclass of `Function`) that overwrites the data array of some variable. There is a demand of implementing in-place activation functions (e.g. relu, sigmoid, tanh) which only requires the output for backprop so that we can discard the input value. Another example is `__setitem__`.

Naive implementation of in-place functions would work well for many use cases, but sometimes it causes an error which is hard to investigate. Consider a function `f` that requires *output* values for backprop and an in-place function `g`. Then, `g(f(x))` corrupts backprop and gradients obtained through this computation might be incorrect. It is because the backprop through `f` requires the original value of `f(x)`, which is overwritten by `g`. [This problem has been already reported in Caffe](https://github.com/BVLC/caffe/issues/2853). The most problematic point is that it does not emit any error and just makes the learning corrupted so that it is hard to investigate the error.

In order to avoid such a situation, we have to let functions declare that they will use the output values in backprop. We then "freeze" the output variables so that applying an in-place function to them emits an error. This is same as how PyTorch resolves this issue (e.g. see [the document on in-place operations](http://pytorch.org/docs/autograd.html#in-place-correctness-checks) and the documentation of Variable following it; EDIT: PyTorch actually does not freeze the varaible, but instead remembers these events and raises an error on backprop).

This solution would break some user codes: we need to make all functions that use the output values in backprop report the future use of the output values (`save_for_backward` of PyTorch) and mark the variable dirty after applying in-place functions.

We have to consider if we should work on it, and if so, we may have to make the API fixed before the release of v2 (otherwise, it would be difficult to make it before the next major release).

**Comment by beam2d** [comment]
See also #2191 for the initial attempts to implement in-place ReLU. Note that this in-place ReLU is unsafe.

**Comment by jekbradbury** [comment]
If you implement the `save_for_backward` of pytorch, your interface will probably be a close enough match to the existing pytorch `autograd.Function` interface that you could try running the pytorch autograd engine. Note that the biggest change is that `save_for_backward` keeps all tensors that you need on the backward pass, meaning that the variables that hold forward tensors shouldn't be a part of the graph itself (since some of them are not actually needed for backward and can be dropped to save memory).

**Comment by beam2d** [comment]
Thank you for the quick follow-up! I'm interested in the detail of `save_for_backward`. I want to confirm if my understanding is correct; do you mean by "the variables that hold forward tensors shouldn't be a part of the graph" that we have to make the graph only consisting from functions that hold any tensors required for backprop and just use Variable.creator as the starting point?

**Comment by jekbradbury** [comment]
Let's say you perform addition during the forward pass, so there's a line like `z = x + y`. Then during the backward pass what needs to happen is basically `x.grad += z.grad; y.grad += z.grad`. Note that `x.data` and `y.data` aren't needed during the backward pass, and the memory they use could have been freed up. There are several kinds of operations like this, including addition/summation and all array ops, in addition to certain in-place activation functions.
In order to implement this significant memory optimization (20-30%+ cut in memory usage depending on the model), the PyTorch developers had to modify the data structure used to store the computation graph. In Chainer the graph consists of `Function` and `Variable` objects, where each `Variable`'s `.creator` holds a reference to the `Function` that created it and each `Function` holds references to all input `Variable`s. In PyTorch the graph consists only of `Function` objects. Specifically, each `Variable` still holds a reference in `.creator` to the `Function` that created it, but each `Function` now holds a single array of `_previous_functions` that contains the `Function`s that were the `.creator`s of each input Variable, allowing the input `Variable`s themselves to be freed by the Python reference counter _unless_ they're marked with `save_for_backward`.
The logic can be found in the .py files in this directory https://github.com/pytorch/pytorch/tree/master/torch/autograd, although those implementations are a little outdated and have been superseded by the C++ implementations in https://github.com/pytorch/pytorch/tree/master/torch/csrc/autograd.

**Comment by beam2d** [comment]
Now I understand the mechanism. It's so smart, I didn't come up with it!

I'm concerning that changing the graph design would break some user codes (esp. those implementing traversing of the graph). Supporting in-place functions alone would be much easier, which is still beneficial to vision applications (e.g. supporting in-place relu may reduce the half of the memory consumption). The PyTorch solution of releasing unneeded forward tensors is much much more effecitve for NLP applications and others requiring complicated computations on top of the autograd layer (e.g. implementing complicated recurrent units without writing custom functions).

**Comment by jekbradbury** [comment]
If you want, you could take the opportunity of a compatibility break and change the graph design in v2 even if you don't actually implement the release of unneeded forward tensors or wrap pytorch's autograd engine.

**Comment by beam2d** [comment]
That's right. I will take a time and possibly have more discussions to proceed this issue. Anyway, thank you for great suggestions!

**Comment by beam2d** [comment]
Close. #2368 should be enough for most cases.
